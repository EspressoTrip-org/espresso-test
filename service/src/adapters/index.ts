export * as policy_cache from './policy-cache';
