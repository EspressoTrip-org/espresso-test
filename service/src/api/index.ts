export * as organization from './organizations';
export * as compilation from './compilation';
export * as policies from './policies';
export * as tokens from './tokens';
export * as roles from './roles';
export * as users from './users';
