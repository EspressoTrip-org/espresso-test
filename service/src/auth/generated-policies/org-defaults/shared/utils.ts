import * as cardinal from '@journeyapps-platform/types-cardinal';
import { createOwnerPolicy } from '../owner';
import { createDeveloperPolicy } from '../developer';

export enum MANAGED_POLICY {
  OWNER = 'owner',
  DEVELOPER = 'developer',
  LOCKED = 'locked'
}

export type ManagedPolicyResponse = Record<MANAGED_POLICY.OWNER | MANAGED_POLICY.DEVELOPER, cardinal.Policy>;

export const createDefaultPoliciesForOrg = (org_id: string): ManagedPolicyResponse => {
  return {
    [MANAGED_POLICY.OWNER]: createOwnerPolicy(org_id),
    [MANAGED_POLICY.DEVELOPER]: createDeveloperPolicy(org_id)
  };
};

export type CreateDefaultRolesPolicyMap = Partial<Record<MANAGED_POLICY, string>>;

export const createDefaultRolesForOrg = (org_id: string, policies: CreateDefaultRolesPolicyMap) => {
  const owner_policy = policies[MANAGED_POLICY.OWNER];
  const owner_role: cardinal.Role = {
    name: cardinal.MANAGED_ROLE.OWNER,
    org_id: org_id,
    policy_ids: [owner_policy!]
  };

  const developer_policy = policies[MANAGED_POLICY.DEVELOPER];
  const developer_role: cardinal.Role = {
    name: cardinal.MANAGED_ROLE.DEVELOPER,
    org_id: org_id,
    policy_ids: [developer_policy!]
  };

  return [owner_role, developer_role];
};
