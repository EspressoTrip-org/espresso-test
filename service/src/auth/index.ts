export * from './generated-policies';
export * from './authorizers';
export * from './context';
export * from './viewer';
