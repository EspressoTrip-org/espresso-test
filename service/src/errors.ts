import * as micro from '@journeyapps-platform/micro';

export class CardinalError extends micro.errors.JourneyError {
  constructor(data: micro.errors.ErrorData) {
    super({
      origin: 'CARDINAL',
      ...data
    });
  }
}

export class Error404 extends CardinalError {
  static CODE = 'ERROR_404';
  constructor(resource: string, id: string) {
    super({
      code: Error404.CODE,
      status: 404,
      description: 'Resource does not exist',
      details: `The resource ${resource}/${id} does not exist`
    });
  }
}

export class ManagedResourceError extends CardinalError {
  static CODE = 'MANAGED_RESOURCE_ERROR';
  constructor(id: string) {
    super({
      code: ManagedResourceError.CODE,
      status: 400,
      description: 'Cannot modify a managed resource',
      details: `The resource ${id} is managed by JourneyApps and cannot be modified directly`
    });
  }
}
