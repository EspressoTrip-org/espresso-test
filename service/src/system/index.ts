import { InterServiceClient } from '@journeyapps-platform/sdk-inter-service';
import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as invalidator from '@journeyapps-platform/recon-event-invalidator';
import * as micro from '@journeyapps-platform/micro';
import * as adapters from '../adapters';
import * as kafka_utils from './kafka';
import * as resolver from '../recon';
import { MongoDB } from './mongo';
import * as kafka from 'kafkajs';
import ENV from '../env';

export type CreateSystemParams = {
  mocks: string[];
  accounts: {
    endpoint: string;
    client_id: string;
    client_secret: string;
  };
  mongo: micro.mongo.MongoDBAdapterOptions;
  kafka: micro.kafka.KafkaConnectionOptions;
  node_env: string;
};

export class System extends micro.system.MicroSystem {
  context: CreateSystemParams;

  mongo: MongoDB;
  kafka: micro.kafka.KafkaClientProvisioner;
  producer: kafka.Producer;
  resolver: entity_resolver.EntityResolver;

  policy_cache: adapters.policy_cache.PolicyCache;

  constructor(params: CreateSystemParams) {
    super();

    this.context = params;

    const inter_service = new InterServiceClient(params.accounts);

    this.mongo = this.withLifecycle(new MongoDB(params.mongo), {
      start(component) {
        return component.connect();
      },
      stop(component) {
        return component.close();
      }
    });

    this.kafka = this.withLifecycle(micro.kafka.createClientProvisioner(params.kafka), {
      start(component) {
        return kafka_utils.ensureTopics({
          kafka: component(),
          production: params.node_env === 'production'
        });
      }
    });

    this.producer = this.withLifecycle(
      this.kafka('producer').producer({
        createPartitioner: kafka.Partitioners.LegacyPartitioner
      }),
      {
        start(component) {
          return component.connect();
        },
        stop(component) {
          return component.disconnect();
        }
      }
    );

    this.resolver = this.withLifecycle(
      resolver.createResolver(
        this.mongo,
        params.mocks.includes('recon')
          ? undefined
          : {
              external: {
                endpoint: params.accounts.endpoint,
                client: inter_service
              }
            }
      ),
      {}
    );

    this.withLifecycle(
      invalidator.createEventInvalidator({
        resolver: this.resolver,
        kafka: this.kafka('invalidator'),
        modules: [invalidator.resource_operation_events_module, invalidator.cardinal_module]
      }),
      {
        start: async (invalidator) => {
          await invalidator.start();
          await this.resolver.init();
        },
        stop(invalidator) {
          return invalidator.stop();
        }
      }
    );

    this.policy_cache = adapters.policy_cache.createInMemoryPolicyCache();
  }

  static fromENV(env: typeof ENV) {
    return new System({
      mocks: env.MOCK,
      accounts: {
        endpoint: env.JWT_ISSUER,
        client_id: env.ACCOUNTS_CLIENT_ID,
        client_secret: env.ACCOUNTS_CLIENT_SECRET
      },
      mongo: {
        uri: env.MONGO_URI,
        database: env.MONGO_DATABASE,
        username: env.MONGO_USERNAME,
        password: env.MONGO_PASSWORD
      },
      kafka: {
        brokers: env.KAFKA_BROKERS,
        username: env.KAFKA_USERNAME,
        password: env.KAFKA_PASSWORD,
        mechanism: env.KAFKA_MECHANISM as any
      },
      node_env: env.NODE_ENV
    });
  }
}
