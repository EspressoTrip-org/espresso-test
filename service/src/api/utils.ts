import { EntityFilter } from '@journeyapps-platform/recon-entity-resolver';
import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import { System } from '../system';
import * as mongo from 'mongodb';
import * as bson from 'bson';
import { Actor } from '@journeyapps-platform/types-cardinal';

export type WithActor<T> = T & {
  actor: Actor;
};

export type OptionalPermissions<T> = T & {
  policies?: cardinal.RawPolicy[];
};

export type RequirePermissions<T> = T & {
  policies: cardinal.RawPolicy[];
};

export const getByIdOrThrow = async <T extends micro.db.Id>(collection: mongo.Collection<T>, id: string) => {
  const res = await collection.findOne(micro.mongo.getById(id) as mongo.Filter<T>);
  if (!res) {
    throw new micro.errors.ResourceNotFound(collection.collectionName, id);
  }
  return micro.mongo.toJson(res);
};

type WrapQueryParams<T> = OptionalPermissions<cardinal.V2ListResourceParams> & {
  query: mongo.Filter<T>;
  model: string;
};
export const wrapQueryWithPermissions = async <T extends micro.db.Id & { org_id?: string; user_id?: string }>(
  system: System,
  params: WrapQueryParams<T>
): Promise<any> => {
  if (!params.policies) {
    return params.query;
  }
  const entity_filter: EntityFilter = {};
  if (params.id && !micro.mongo.isQueryFilter(params.id)) {
    entity_filter.id = params.id;
  }
  if (params.org_id && !micro.mongo.isQueryFilter(params.org_id)) {
    entity_filter.organization_id = params.org_id;
  }
  if (params.user_id && !micro.mongo.isQueryFilter(params.user_id)) {
    entity_filter.user_id = params.user_id;
  }

  const ids = await system.resolver.resolveResourceIds({
    policies: params.policies,
    model: params.model,
    entity_filter
  });

  return {
    $and: [
      {
        _id: {
          $in: ids.map((id) => new bson.ObjectId(id))
        }
      },
      params.query
    ]
  };
};
