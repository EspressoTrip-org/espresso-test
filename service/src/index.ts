import '@journeyapps-platform/micro/register';

import * as micro from '@journeyapps-platform/micro';
import * as routes from './router/routes/index';
import { Router } from './router/router';
import * as consumers from './consumers/index';
import { System } from './system/index';
import * as auth from './auth/index';
import fastify from 'fastify';
import env from './env';

const server = fastify();

const system = System.fromENV(env);
const jwt_decoder = auth.createJWTDecoder(system, {
  issuer_url: env.JWT_ISSUER,
  verify: !env.DEV_MODE_DO_NOT_ENABLE_IN_PRODUCTION_OR_YOU_WILL_BE_FIRED
});

const token_decoder = auth.createTokenDecoder(system);
const dev_decoder = auth.createDevDecoder();

const decoders = [jwt_decoder, token_decoder];
if (env.DEV_MODE_DO_NOT_ENABLE_IN_PRODUCTION_OR_YOU_WILL_BE_FIRED) {
  decoders.push(dev_decoder);
}

server.register(
  Router.plugin({
    routes: [...routes.v1_routes, ...routes.v2_routes, ...micro.router.createProbeRoutes()],
    contextProvider: async (payload) => {
      return {
        system: system,
        viewer: await auth.createViewer({
          token: payload.request.headers['authorization'] || '',
          system,
          decoders
        }),
        sha: env.SHA
      };
    }
  })
);

(async () => {
  await system.start();

  await micro.fastify.startServer(server, env.PORT);
  micro.logger.info(`HTTP server running on port ${env.PORT}`);
  await micro.signals.getSystemProbe().ready();

  const [resource_events_consumer, cache_invalidator] = await Promise.all([
    consumers.createResourceOperationEventsConsumer(system),
    consumers.createCacheInvalidator(system)
  ]);

  micro.signals.getTerminationHandler()?.handleTerminationSignal(async () => {
    await Promise.all([resource_events_consumer.disconnect(), cache_invalidator.disconnect()]);
  });
})().catch((err) => {
  micro.logger.error('Something went wrong while booting', err);
  process.exit(1);
});
