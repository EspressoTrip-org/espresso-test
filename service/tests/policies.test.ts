import { beforeAll, afterEach, afterAll, describe, test, expect } from 'vitest';

import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as errors from '../src/errors';
import { System } from '../src/system';
import * as auth from '../src/auth';
import * as api from '../src/api';
import * as mocks from './mocks';
import * as utils from './utils';
import * as bson from 'bson';
import * as _ from 'lodash';
import { CardinalModel, PolicyModelAction } from '@journeyapps-platform/cardinal-catalog';

let system: System;

beforeAll(async () => {
  system = await mocks.createMockedSystem();
  await system.start();
});

afterEach(() => {
  (system.producer as any).clear();
});

afterAll(async () => {
  await system.stop();
});

describe('policies', () => {
  describe('policy-crud', () => {
    test('should list policies by scope', async () => {
      await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Scoped Policy',
          statements: [],
          org_id: '123'
        }
      });

      await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Unscoped Policy',
          statements: []
        }
      });

      const scoped = await api.policies.listScopedPoliciesForOrganization(system.mongo, {
        org_id: '123'
      });
      const unscoped = await api.policies.listUnscopedPolicies(system.mongo);

      expect(utils.removeVolatile(scoped)).toMatchSnapshot();
      expect(utils.removeVolatile(unscoped)).toMatchSnapshot();
    });

    test('should successfully create a scoped policy', async () => {
      const res = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Some policy name',
          description: 'Some policy description',
          statements: [
            {
              actions: ['create', 'read'],
              resources: [
                {
                  scope: '*',
                  selector: {
                    id: '*',
                    model: '*'
                  }
                }
              ]
            }
          ],
          org_id: '123'
        }
      });

      expect(utils.removeVolatile(res)).toMatchSnapshot();
      expect(utils.getEvents(system)).toMatchSnapshot();
    });

    test('should successfully update a policy', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          org_id: '123',
          statements: [],
          name: 'policy'
        }
      });

      await api.policies.updatePolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          ...policy,
          name: 'Updated Policy'
        }
      });

      const updated_policy = await api.policies.getPolicy(system.mongo, policy.id);
      expect(utils.removeVolatile(updated_policy)).toMatchSnapshot();

      expect(utils.getEvents(system)).toMatchSnapshot();
    });

    test('should remove associations from related resources on delete', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          org_id: '123',
          statements: [],
          name: 'policy'
        }
      });

      const now = new Date();

      const remaining_policy_id = '5e6f5519ede3631176831f4c';
      await system.mongo.policies.insertOne({
        _id: new bson.ObjectId(remaining_policy_id),
        updated_at: now,
        created_at: now,
        name: 'dangling',
        org_id: '123',
        statements: []
      });

      const user = await system.mongo.users.insertOne({
        _id: new bson.ObjectId(),

        org_id: '123',
        email: 'test@journeyapps.com',
        policy_ids: [policy.id, remaining_policy_id]
      });

      const role = await system.mongo.roles.insertOne({
        _id: new bson.ObjectId(),

        name: 'test-role',
        org_id: '123',
        policy_ids: [policy.id, remaining_policy_id],
        created_at: now,
        updated_at: now
      });

      const token = await system.mongo.tokens.insertOne({
        _id: new bson.ObjectId(),

        description: 'test token',
        value: 'nothing',
        org_id: '123',
        policy_ids: [policy.id, remaining_policy_id],
        created_at: now,
        updated_at: now
      });

      await api.policies.deletePolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        id: policy.id
      });

      const updated_user = await system.mongo.users.findOne({
        _id: user.insertedId
      });
      const updated_role = await system.mongo.roles.findOne({
        _id: role.insertedId
      });
      const updated_token = await system.mongo.tokens.findOne({
        _id: token.insertedId
      });

      expect(updated_user?.policy_ids?.length).toBe(1);
      expect(updated_user?.policy_ids?.[0]).toBe(remaining_policy_id);

      expect(updated_role?.policy_ids?.length).toBe(1);
      expect(updated_role?.policy_ids?.[0]).toBe(remaining_policy_id);

      expect(updated_token?.policy_ids?.length).toBe(1);
      expect(updated_token?.policy_ids?.[0]).toBe(remaining_policy_id);

      expect(utils.getEvents(system)).toMatchSnapshot();
    });
  });

  describe('policy-auth', () => {
    const unscoped_policy: cardinal.Policy = {
      name: 'scoped-policy',
      statements: []
    };

    const scoped_policy: cardinal.Policy = {
      name: 'scoped-policy',
      org_id: '123',
      statements: []
    };

    const unscoped_create_permission: cardinal.Policy = {
      statements: [
        {
          actions: ['create'],
          resources: [
            {
              selector: {
                model: CardinalModel.POLICY,
                id: '*'
              }
            }
          ]
        }
      ]
    };

    const scoped_create_permission: cardinal.Policy = _.merge({}, unscoped_create_permission, {
      statements: [
        {
          resources: [
            {
              scope: '123'
            }
          ]
        }
      ]
    });

    const unscoped_read_permission: cardinal.Policy = {
      statements: [
        {
          actions: ['read'],
          resources: [
            {
              selector: {
                model: 'app',
                id: '*'
              }
            }
          ]
        }
      ]
    };

    const scoped_read_permission: cardinal.Policy = _.merge({}, unscoped_read_permission, {
      statements: [
        {
          resources: [
            {
              scope: '*'
            }
          ]
        }
      ]
    });

    const unscoped_statement: cardinal.PolicyStatement = {
      actions: ['read'],
      resources: [
        {
          selector: {
            model: 'app',
            id: '*'
          }
        }
      ]
    };

    const scoped_statement: cardinal.PolicyStatement = {
      actions: ['read'],
      resources: [
        {
          scope: '123',
          selector: {
            model: 'app',
            id: '*'
          }
        }
      ]
    };

    test('should be unauthorized to create policy [operation]', async () => {
      const context = mocks.createMockedContext(system, []);

      const res = await auth.policies.authorizedToCreatePolicy({ params: scoped_policy, context });
      expect(res.authorized).toBeFalsy();
    });

    test('should be unauthorized to create policy [scope]', async () => {
      const scoped_context = mocks.createMockedContext(system, [scoped_create_permission]);
      const unscoped_context = mocks.createMockedContext(system, [unscoped_create_permission]);

      const res1 = await auth.policies.authorizedToCreatePolicy({ params: unscoped_policy, context: scoped_context });
      expect(res1.authorized).toBeFalsy();

      const res2 = await auth.policies.authorizedToCreatePolicy({ params: scoped_policy, context: unscoped_context });
      expect(res2.authorized).toBeFalsy();
    });

    test('should be unauthorized to create policy [policy/statement scope mismatch]', async () => {
      const scoped_context = mocks.createMockedContext(system, [
        scoped_create_permission,
        {
          statements: [
            {
              actions: ['read'],
              resources: [
                {
                  scope: '12345',
                  selector: {
                    model: 'app',
                    id: '*'
                  }
                }
              ]
            }
          ]
        }
      ]);

      // This should be unauthorized because the statement scopes do not match the resource org_id
      const res = await auth.policies.authorizedToCreatePolicy({
        params: {
          ...scoped_policy,
          statements: [
            {
              actions: ['read'],
              resources: [
                {
                  scope: '12345',
                  selector: {
                    model: 'app',
                    id: '*'
                  }
                }
              ]
            }
          ]
        },
        context: scoped_context
      });
      expect(res.authorized).toBeFalsy();
    });

    test('should be authorized to create policy [operation]', async () => {
      const scoped_context = mocks.createMockedContext(system, [scoped_create_permission]);
      const unscoped_context = mocks.createMockedContext(system, [unscoped_create_permission]);

      const res1 = await auth.policies.authorizedToCreatePolicy({ params: scoped_policy, context: scoped_context });
      expect(res1.authorized).toBeTruthy();

      const res2 = await auth.policies.authorizedToCreatePolicy({ params: unscoped_policy, context: unscoped_context });
      expect(res2.authorized).toBeTruthy();
    });

    test('should be unauthorized to create policies [statements 1]', async () => {
      const policy: cardinal.Policy = {
        ...scoped_policy,
        statements: [scoped_statement]
      };

      const scoped_context = mocks.createMockedContext(system, [scoped_create_permission]);

      const res = await auth.policies.authorizedToCreatePolicy({ params: policy, context: scoped_context });
      expect(res.authorized).toBeFalsy();
    });

    test('should be unauthorized to create policies [statements 2]', async () => {
      const new_scoped_policy: cardinal.Policy = {
        ...scoped_policy,
        statements: [
          {
            actions: ['read'],
            resources: [
              {
                scope: '123',
                selector: {
                  model: 'app',
                  id: '*'
                }
              }
            ]
          }
        ]
      };

      const scoped_context = mocks.createMockedContext(system, [
        scoped_create_permission,
        {
          statements: [
            {
              actions: ['read'],
              resources: [
                {
                  scope: '123',
                  selector: {
                    model: 'app',
                    id: '123'
                  }
                }
              ]
            }
          ]
        }
      ]);

      const res1 = await auth.policies.authorizedToCreatePolicy({ params: new_scoped_policy, context: scoped_context });
      expect(res1.authorized).toBeFalsy();
    });

    test('should be authorized to create policies [statements]', async () => {
      const new_scoped_policy: cardinal.Policy = {
        ...scoped_policy,
        statements: [scoped_statement]
      };

      const new_unscoped_policy: cardinal.Policy = {
        ...unscoped_policy,
        statements: [unscoped_statement]
      };

      const scoped_context = mocks.createMockedContext(system, [scoped_create_permission, scoped_read_permission]);
      const unscoped_context = mocks.createMockedContext(system, [
        unscoped_create_permission,
        unscoped_read_permission
      ]);

      const res1 = await auth.policies.authorizedToCreatePolicy({ params: new_scoped_policy, context: scoped_context });
      expect(res1.authorized).toBeTruthy();

      const res2 = await auth.policies.authorizedToCreatePolicy({
        params: new_unscoped_policy,
        context: unscoped_context
      });
      expect(res2.authorized).toBeTruthy();
    });

    test('should be unauthorized to create policies with $or values [statements]', async () => {
      const current_policy: cardinal.Policy = {
        statements: [
          {
            actions: ['read'],
            resources: [
              {
                scope: '123',
                selector: {
                  model: 'app',
                  id: '123'
                }
              }
            ]
          }
        ]
      };

      // Here we have an intersection of 'read' with current assigned policies. This should still not
      // allow creation due to the additional 'create' action also present
      const action_intersection_policy: cardinal.Policy = {
        ...scoped_policy,
        statements: [
          {
            actions: ['read', 'create'],
            resources: [
              {
                scope: '123',
                selector: {
                  model: ['app'],
                  id: '123'
                }
              }
            ]
          }
        ]
      };

      // Here we have an intersection of 'app' with current assigned policies. This should still not
      // allow creation due to the additional 'user' model also present
      const model_intersection_policy: cardinal.Policy = {
        ...scoped_policy,
        statements: [
          {
            actions: ['read'],
            resources: [
              {
                scope: '123',
                selector: {
                  model: ['app', 'user'],
                  id: '123'
                }
              }
            ]
          }
        ]
      };

      const context = mocks.createMockedContext(system, [scoped_create_permission, current_policy]);

      const res1 = await auth.policies.authorizedToCreatePolicy({
        params: action_intersection_policy,
        context: context
      });
      expect(res1.authorized).toBeFalsy();
      const res2 = await auth.policies.authorizedToCreatePolicy({
        params: model_intersection_policy,
        context: context
      });
      expect(res2.authorized).toBeFalsy();
    });

    test('should be unauthorized to modify policy [managed]', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          ...scoped_policy,
          managed: true,
          statements: []
        }
      });

      const scoped_context = mocks.createMockedContext(system, [
        {
          statements: [
            {
              actions: ['write'],
              resources: [
                {
                  scope: '123',
                  selector: {
                    model: CardinalModel.POLICY,
                    id: '*'
                  }
                }
              ]
            },
            {
              actions: [PolicyModelAction.DELETE],
              resources: [
                {
                  scope: '123',
                  selector: {
                    model: CardinalModel.POLICY,
                    id: '*'
                  }
                }
              ]
            }
          ]
        }
      ]);

      await expect(auth.policies.authorizedToUpdatePolicy({ params: policy, context: scoped_context })).rejects.toThrow(
        errors.ManagedResourceError
      );

      await expect(auth.policies.authorizedToDeletePolicy({ params: policy, context: scoped_context })).rejects.toThrow(
        errors.ManagedResourceError
      );
    });

    test('should only return policies authorized to read', async () => {
      const org_id_1 = '5f8ec0f7dbfb68e57f3bfa23';
      const org_id_2 = '5f8ec103dbfb68e57f3bfa25';

      const p1 = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Scoped Policy 1',
          statements: [],
          org_id: org_id_1
        }
      });
      await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Scoped Policy 2',
          statements: [],
          org_id: org_id_1
        }
      });
      await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Scoped Policy 3',
          statements: [],
          org_id: org_id_2
        }
      });

      await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          name: 'Unscoped Policy',
          statements: []
        }
      });

      await system.resolver.init();

      const permissions_1: cardinal.Policy[] = [
        {
          statements: [
            {
              actions: ['read'],
              resources: [
                {
                  scope: org_id_1,
                  selector: {
                    model: CardinalModel.POLICY,
                    id: p1.id
                  }
                }
              ]
            }
          ]
        }
      ];

      const permissions_2: cardinal.Policy[] = [
        {
          statements: [
            {
              actions: ['read'],
              resources: [
                {
                  scope: org_id_1,
                  selector: {
                    model: CardinalModel.POLICY,
                    id: '*'
                  }
                }
              ]
            }
          ]
        }
      ];

      const scoped_1 = await api.policies.listPolicies(system, {
        org_id: org_id_1,
        policies: permissions_1
      });
      const scoped_2 = await api.policies.listPolicies(system, {
        org_id: org_id_1,
        policies: permissions_2
      });
      const scoped_3 = await api.policies.listPolicies(system, {
        org_id: org_id_2,
        policies: permissions_1
      });
      const scoped_4 = await api.policies.listPolicies(system, {
        org_id: org_id_1,
        policies: []
      });
      const unscoped = await api.policies.listPolicies(system, {
        org_id: {
          exists: false
        },
        policies: permissions_1
      });

      expect(scoped_1.items.length).toEqual(1);
      expect(scoped_1.items[0].id).toEqual(p1.id);

      expect(scoped_2.items.length).toEqual(2);
      expect(utils.removeVolatile(scoped_2.items)).toMatchSnapshot();

      expect(scoped_3.items).toEqual([]);
      expect(scoped_4.items).toEqual([]);
      expect(unscoped.items).toEqual([]);
    });
  });
});
