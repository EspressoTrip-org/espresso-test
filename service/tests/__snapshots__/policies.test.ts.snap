// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`policies > policy-auth > should only return policies authorized to read 1`] = `
[
  {
    "name": "Scoped Policy 2",
    "org_id": "5f8ec0f7dbfb68e57f3bfa23",
    "statements": [],
  },
  {
    "name": "Scoped Policy 1",
    "org_id": "5f8ec0f7dbfb68e57f3bfa23",
    "statements": [],
  },
]
`;

exports[`policies > policy-crud > should list policies by scope 1`] = `
[
  {
    "name": "Scoped Policy",
    "org_id": "123",
    "statements": [],
  },
]
`;

exports[`policies > policy-crud > should list policies by scope 2`] = `
[
  {
    "name": "Unscoped Policy",
    "statements": [],
  },
]
`;

exports[`policies > policy-crud > should remove associations from related resources on delete 1`] = `
[
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "policy",
      "org_id": "123",
      "statements": [],
    },
    "type": "AUTH.POLICY_CREATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "email": "test@journeyapps.com",
      "org_id": "123",
      "policy_ids": [
        "5e6f5519ede3631176831f4c",
      ],
    },
    "type": "AUTH.USER_ASSIGNMENTS_CHANGED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "description": "test token",
      "org_id": "123",
      "policy_ids": [
        "5e6f5519ede3631176831f4c",
      ],
    },
    "type": "AUTH.TOKEN_UPDATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "test-role",
      "org_id": "123",
      "policy_ids": [
        "5e6f5519ede3631176831f4c",
      ],
    },
    "type": "AUTH.ROLE_UPDATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "policy",
      "org_id": "123",
      "statements": [],
    },
    "type": "AUTH.POLICY_DELETED",
  },
]
`;

exports[`policies > policy-crud > should successfully create a scoped policy 1`] = `
{
  "description": "Some policy description",
  "name": "Some policy name",
  "org_id": "123",
  "statements": [
    {
      "actions": [
        "create",
        "read",
      ],
      "resources": [
        {
          "scope": "*",
          "selector": {
            "id": "*",
            "model": "*",
          },
        },
      ],
    },
  ],
}
`;

exports[`policies > policy-crud > should successfully create a scoped policy 2`] = `
[
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "description": "Some policy description",
      "name": "Some policy name",
      "org_id": "123",
      "statements": [
        {
          "actions": [
            "create",
            "read",
          ],
          "resources": [
            {
              "scope": "*",
              "selector": {
                "id": "*",
                "model": "*",
              },
            },
          ],
        },
      ],
    },
    "type": "AUTH.POLICY_CREATED",
  },
]
`;

exports[`policies > policy-crud > should successfully update a policy 1`] = `
{
  "name": "Updated Policy",
  "org_id": "123",
  "statements": [],
}
`;

exports[`policies > policy-crud > should successfully update a policy 2`] = `
[
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "policy",
      "org_id": "123",
      "statements": [],
    },
    "type": "AUTH.POLICY_CREATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "Updated Policy",
      "org_id": "123",
      "statements": [],
    },
    "type": "AUTH.POLICY_UPDATED",
  },
]
`;
